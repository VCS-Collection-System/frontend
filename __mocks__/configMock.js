export const getConfig = (URL, configUrl, config) => {
  if(URL.indexOf(configUrl) > -1 ) {
    const country = URL.substring(configUrl.length);
    const countryConfig = config[country] || null;
    return Promise.resolve({data: countryConfig});
  } else {
    return Promise.reject({data: null});
  }
}