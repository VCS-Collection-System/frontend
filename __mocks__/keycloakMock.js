module.exports = {
    token: "bogustoken", 
    authenticated: false,
    loadUserProfile: jest.fn( ()=> {return Promise.resolve({
        attributes: {
            workforceid: ["mockId"]
        }
    })} ),
    login: jest.fn(),
    logout: jest.fn(),
    hasRealmRole: jest.fn(),
    profile: {
        firstName: "mockFirst",
        lastName: "mockLast",
        email: "mock@email.com",
        attributes: {
            workforceid: ["mockId"],
            agency: ["USA"]
        }
    }
};
