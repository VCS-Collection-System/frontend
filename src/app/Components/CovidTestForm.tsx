import React, { useContext, useEffect, useState } from "react";
import { 
  Checkbox,
  Form, 
  FormGroup, 
  TextInput, 
  FormSelect, 
  FormSelectOption, 
  DatePicker, 
  FileUpload, 
  Button, 
  Flex,
  FlexItem,
  Text
} from "@patternfly/react-core";
import { useKeycloak } from "@react-keycloak/web";
import { CheckCircleIcon, OutlinedPaperPlaneIcon } from "@patternfly/react-icons";
import { Controller, useForm } from "react-hook-form";
import { useHistory } from "react-router-dom";
import Feature, { showFeature } from "@app/Components/Feature";
import { ConsentContext } from "@app/context";
import { 
  dateConvert, 
  americanDateFormat,
  covidTestDateRules,
  covidTestDateValidator,
  birthDateValidator,
  birthDateRules
} from "@app/utils/utils";
import {
  ALTERNATE_CONSENT_LABEL
} from "@app/utils/constants";

const CovidTestForm = (props) => {

  const history = useHistory();
  const { keycloak } = useKeycloak();
  const { control, handleSubmit } = useForm();
  const [isTestFileRejected, setIsTestFileRejected] = useState(false);
  const { setConsented } = useContext(ConsentContext);
  const [ isCovidTest, setIsCovidTest ] = useState(false); 
  const [ isRecovery, setIsRecovery ] = useState(false);
  const [allowPDF, setAllowPDF] = useState(false);

  let resultLabel = "";
  if(isCovidTest) {
    resultLabel += "COVID Test Result";
    if(isRecovery) {
      resultLabel += " or Certificate of Recovery";
    }
  } else {
    resultLabel += "Type of Submission";
  }
  resultLabel += ".";

  let resultHelper = "Please select the ";
  if(isCovidTest) {
    resultHelper += "COVID test result";
    if(isRecovery) {
      resultHelper += " or certificate of recovery";
    }
  } else {
    resultHelper += "type of submission";
  }
  resultHelper += ".";

  let dateHelper = "Please enter the date the ";
  if(isCovidTest) {
    dateHelper += "COVID test";
    if(isRecovery) {
      dateHelper += " or certificate of recovery";
    }
  } else {
    dateHelper += "certificate of recovery";
  }
  dateHelper += " was administered.";

  useEffect(() => {
    showFeature("covidTest", keycloak, props.employeeCountry).then(config => {
      setIsCovidTest(config);
    });

    showFeature("recovery", keycloak, props.employeeCountry).then(config => {
      setIsRecovery(config);
    });

    showFeature("allowPdf", keycloak, props.employeeCountry).then(config => {
      setAllowPDF(config);
    });
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const ImageHelperText = ({isInvalidText}) => {    
    return (
      <Text className={isInvalidText ? "helperText warningText" : "helperText"}> 
        Must be a {allowPDF ? "PDF," : ""} JPG, JPEG, or PNG file no larger than 3.9 MB 
        <br />
        Please refer to &nbsp;
        <a href="https://redhat.service-now.com/help?id=kb_article_view&sysparm_article=KB0015729" target="_blank"  rel="noreferrer">
          Image Submission FAQ
        </a> 
        &nbsp; for further guidance. 
      </Text>
    )
  }

  const getAcceptableProofs = () => {
    return ".jpg,.jpeg,.png" + (allowPDF ? ",.pdf" : "" );
  }

  return (
    <Form onSubmit={handleSubmit(props.onSubmit)}>
      {!props.isHRSubmit ? (
        <Feature name="consentCheckboxAlternate">
          <Controller
            name="consent"
            control={control}
            render={({ field, fieldState }) => (
              <FormGroup 
                fieldId="consent-checkbox" 
                isRequired
                label="Consent Agreement"
                helperTextInvalid="This field is required, please complete consent agreement."
                validated={fieldState.error ? "error" : "default"}
              >
                <Checkbox 
                  label={ALTERNATE_CONSENT_LABEL} 
                  aria-label="consent-checkbox" 
                  id="consent-checkbox" 
                  isChecked={field.value}
                  onChange={(value) => {
                    setConsented(value);
                    field.onChange(value);
                  }}
                />
              </FormGroup>
            )}
            rules={{ required: true }}
          />
        </Feature>
      ) : ""}
      <FormGroup
        fieldId="employee-full-name"
        helperText={
          props.isHRSubmit ? 
          "Make sure this matches the full name of the person you are submitting for." : 
          "Make sure this matches your full name."
        }
        label="Full Name"
      >
        <TextInput
          aria-label="employee-full-name"
          isDisabled
          type="text"
          value={props.employeeName}
        />
      </FormGroup>
      <Controller
        name="alt_first_name"
        control={control}
        defaultValue={props.isHRSubmit ? "" : (sessionStorage.getItem("alt_first_name") || "")}
        render={({ field, fieldState }) => (
          <FormGroup 
            fieldId="alt-first-name" 
            helperText={
              props.isHRSubmit ? 
              "Individual's first name as displayed on the documentation if different than above." : 
              "Your first name as displayed on the documentation if different than above."
            }
            label="Preferred First Name (if applicable)"
            validated={fieldState.error ? "error" : "default"}
          >
            <TextInput
              aria-label="alt-first-name"
              type="text"
              value={field.value}
              onChange={field.onChange}
            />
          </FormGroup>
        )}
      />
      <Controller
        name="alt_last_name"
        control={control}
        defaultValue={props.isHRSubmit ? "" : (sessionStorage.getItem("alt_last_name") || "")}
        render={({ field, fieldState }) => (
          <FormGroup 
            fieldId="alt-last-name" 
            helperText={
              props.isHRSubmit ? 
              "Individual's last name as displayed on the documentation if different than above." : 
              "Your last name as displayed on the documentation if different than above."
            }
            label="Preferred Last Name (if applicable)"
            validated={fieldState.error ? "error" : "default"}
          >
            <TextInput
              aria-label="alt-last-name"
              type="text"
              value={field.value}
              onChange={field.onChange}
            />
          </FormGroup>
        )}
      />
      <Feature name="dateOfBirth" employeeCountry={props.employeeCountry}>
        <Controller
          name="date_of_birth"
          control={control}
          defaultValue={props.isHRSubmit ? "" : (dateConvert(sessionStorage.getItem("date_of_birth") || ""))}
          render={({ field, fieldState }) => (
            <FormGroup
              helperText={
                props.isHRSubmit ?
                "Please enter the birth date as shown on the documentation." :
                "Please enter your birth date."
              }
              helperTextInvalid="This field is required, please enter a valid date."
              fieldId="date_of_birth"
              label="Date of Birth"
              isRequired
              validated={fieldState.error ? "error" : "default"}
            >
              <DatePicker
                aria-label="date_of_birth"
                onChange={field.onChange}
                dateFormat={date => date.toLocaleDateString('en-US', { year: 'numeric', month: '2-digit', day: '2-digit' }).replace(/\//g, '-')}
                dateParse={americanDateFormat}
                placeholder="MM-DD-YYYY"
                value={field.value}
                validators={[birthDateValidator]}
              />
            </FormGroup>
          )}
          rules={{ required: true, validate: birthDateRules, pattern: /^(0[1-9]|1[0-2])\-(0[1-9]|1\d|2\d|3[01])\-(19|20)\d{2}$/ }}
        />
      </Feature>
      <Controller
        name="test_results"
        control={control}
        rules={{ required: true }}
        defaultValue={props.isHRSubmit ? "" : (sessionStorage.getItem("test_results") || "")}
        render={({ field, fieldState }) => (
          <FormGroup
            fieldId="test-results-input"
            helperText={resultHelper}
            isRequired
            validated={fieldState.error ? "error" : "default"}
            label={resultLabel}
            helperTextInvalid="This field is required to proceed."
          >
            <FormSelect
              aria-label="test-results-input"
              onChange={field.onChange}
              validated={fieldState.error ? "error" : "default"}
              value={field.value}
            >
              <FormSelectOption
                isPlaceholder
                isDisabled
                label="Please Select"
              />
              <Feature name="covidTest" employeeCountry={props.employeeCountry}>
                <FormSelectOption label="Negative" value="Negative" />
                <FormSelectOption label="Positive" value="Positive" />
                <FormSelectOption label="Inconclusive" value="Inconclusive" />
              </Feature>
              <Feature name="recovery" employeeCountry={props.employeeCountry}>
                <FormSelectOption label="Certificate of Recovery" value="Recovery" />
              </Feature>
            </FormSelect>
          </FormGroup>
        )}
      />
      <Controller
        control={control}
        defaultValue={props.isHRSubmit ? "" : (dateConvert(sessionStorage.getItem("test_date") || ""))}
        name="test_date"
        render={({ field, fieldState }) => (
          <FormGroup
            label="Date"
            fieldId="test-results-date"
            helperText={dateHelper}
            helperTextInvalid="This field is required to proceed."
            isRequired
            validated={fieldState.error ? "error" : "default"}
          >
            <DatePicker
              aria-label="test-results-date"
              dateFormat={date => date.toLocaleDateString('en-US', { year: 'numeric', month: '2-digit', day: '2-digit' }).replace(/\//g, '-')}
              dateParse={americanDateFormat}
              placeholder="MM-DD-YYYY"
              onChange={field.onChange}
              value={field.value}
              validators={[covidTestDateValidator]}
            />
          </FormGroup>
        )}
        rules={{ required: true, validate: covidTestDateRules, pattern: /^(0[1-9]|1[0-2])\-(0[1-9]|1\d|2\d|3[01])\-(19|20)\d{2}$/  }}
      />
      {!props.isHRSubmit &&
        <Controller
          control={control}
          defaultValue={props.onLoad}
          name="test_image_file"
          render={({ field, fieldState }) => (
            <FormGroup
              fieldId="test-record-image"
              isRequired
              label="Provide a copy of supporting documentation."
              helperText={<ImageHelperText isInvalidText={false}/>}
              helperTextInvalid={<ImageHelperText isInvalidText={true}/>}
              validated={isTestFileRejected || fieldState.error ? "error" : "default"}
            >
              <FileUpload
                hideDefaultPreview
                filename={field.value ? field.value["name"] : ""}
                id="test-record-image"
                value={field.value}
                onChange={field.onChange}
                dropzoneProps={{
                  accept: getAcceptableProofs(),
                  maxSize: 3900000,
                  onDropRejected: () => {
                    setIsTestFileRejected(true);
                  },
                  onDropAccepted: () => {
                    setIsTestFileRejected(false);
                  }
                }}
                validated={isTestFileRejected ? "error" : "default"}
              />
            </FormGroup>
          )}
          rules={{ required: true }}
        />
      }
      {props.isHRSubmit &&
        <Controller
          name="alternate_email"
          defaultValue={""}
          control={control}
          render={({ field, fieldState }) => (
            <FormGroup
              id="alternative-email"
              label="Alternate Email (optional)"
              helperText="Optional email for notifications"
              helperTextInvalid="This needs to be a valid email address."
              fieldId="alternative-email"
              validated={fieldState.error ? "error" : "default"}
            >
              <TextInput
                aria-label="alternative-email"
                onChange={field.onChange}
                placeholder="example@mail.com"
                type="email"
                value={field.value}
              />
            </FormGroup>
          )}
          rules={{
            required: false,
            pattern: /[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,3}/g, // NOSONAR
          }}
        />
      }
      {props.isHRSubmit ? 
        <FormGroup fieldId="accept-test-buttons">
          <Flex>
            <FlexItem align={{ default: "alignRight" }}>
              <Button
                icon={<OutlinedPaperPlaneIcon />}
                variant="primary"
                type="submit"
                isDisabled={props.isFormSubmitting}
                isLoading={props.isFormSubmitting}
              >
                Submit
              </Button>
            </FlexItem>
          </Flex>
        </FormGroup>
        :
        <FormGroup fieldId="accept-test-buttons">
          <Button 
            icon={<CheckCircleIcon />} 
            variant="primary" 
            type="submit"
          >
            Accept
          </Button>
          <Button 
            variant="plain" 
            onClick={() => history.push("/attestmenu")}
          >
            Back
          </Button>
        </FormGroup>
      }
    </Form>
  )
};

export default CovidTestForm;
