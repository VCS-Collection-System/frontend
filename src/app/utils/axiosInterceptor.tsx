import axios from "axios";

const axiosInterceptor = axios.create({})

axiosInterceptor.interceptors.request.use(
    config => {
        config.headers['Authorization'] = 'Bearer ' + sessionStorage.getItem("token");
        return config;
    },
    error => {
        Promise.reject(error)
    });
export default axiosInterceptor;