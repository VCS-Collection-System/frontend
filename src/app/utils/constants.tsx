declare global {  
  interface Window {
    REACT_APP_PAM_URL_BASE: string;
    REACT_APP_PAM_CONTAINER_ID: string;
    REACT_APP_KEYCLOAK_COUNTRY: string;
    REACT_APP_KEYCLOAK_EMAIL: string;
    REACT_APP_KEYCLOAK_FIRST_NAME: string;
    REACT_APP_KEYCLOAK_ID: string;
    REACT_APP_KEYCLOAK_LAST_NAME: string;
    REACT_APP_MAINTENANCE_BANNER: boolean;
    REACT_APP_MAINTENANCE_BANNER_TEXT: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_FEATURE_FLAGS: Array<Record<string, any>>;
    REACT_APP_MAX_DOSES: Record<string, number>;
    REACT_APP_CONSENT_LABEL: string;
    REACT_APP_ALTERNATE_CONSENT_LABEL: string;
    REACT_APP_BLOCKED_COUNTRIES: Array<string>;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_GLOBAL_CONFIG: Record<string, any>;
  }
}

// PAM urls
export const PAM_BASE_URL = window.REACT_APP_PAM_URL_BASE; 
export const SUBMIT_VAX_URL = PAM_BASE_URL + '/attestation/vax';
export const SUBMIT_TEST_URL = PAM_BASE_URL + '/attestation/covid-test-result';
export const PAM_TASK_URL = PAM_BASE_URL + '/tasks/';
export const RESERVED_TASK_URL = PAM_TASK_URL + 'owners?status=Reserved&user=';
export const INPROGRESS_TASK_URL = PAM_TASK_URL + 'owners?status=InProgress&user=';
export const ALL_TASK_URL = PAM_TASK_URL + 'owners?status=Ready&pageSize=200';
export const PRIORITY_URL = PAM_BASE_URL + '/query/priority-results';
export const PRIORITY_CONFIRM_URL = PAM_BASE_URL + '/query/priority-result/';
export const ATTACHMENT_URL = PAM_BASE_URL + '/attestation/attachment';
export const EMPLOYEE_URL = PAM_BASE_URL + '/query/employee/';
export const EMPLOYEE_NAME_URL = PAM_BASE_URL + '/query/employee-name/';
export const EMPLOYEE_VAX_HISTORY_URL = PAM_BASE_URL + '/query/vax-document/';
export const EMPLOYEE_TEST_HISTORY_URL = PAM_BASE_URL + '/query/test-result-document/';
export const VAX_TASK_DOCUMENT_URL = PAM_BASE_URL + '/query/task/document/';
export const DELETE_VAX_URL = PAM_BASE_URL + '/attestation/delete-vax-record/';
export const CONFIG_URL = PAM_BASE_URL + '/query/country-configs/';
export const SAVE_CONFIG_URL = PAM_BASE_URL + '/attestation/country-configs';

// List of countries allow to use this app
export const BLOCKED_COUNTRIES = window.REACT_APP_BLOCKED_COUNTRIES || []

export const KONAMI_CODE = [ "ArrowUp", "ArrowUp", "ArrowDown", "ArrowDown", "ArrowLeft", "ArrowRight", "ArrowLeft", "ArrowRight", "KeyB", "KeyA"];

// Maintenance Banner
export const MAINTENANCE_BANNER = window.REACT_APP_MAINTENANCE_BANNER || false
export const MAINTENANCE_BANNER_TEXT = window.REACT_APP_MAINTENANCE_BANNER_TEXT || ""

// Keycloak claims
export const KEYCLOAK_COUNTRY = window.REACT_APP_KEYCLOAK_COUNTRY
export const KEYCLOAK_EMAIL = window.REACT_APP_KEYCLOAK_EMAIL
export const KEYCLOAK_FIRST_NAME = window.REACT_APP_KEYCLOAK_FIRST_NAME
export const KEYCLOAK_ID = window.REACT_APP_KEYCLOAK_ID
export const KEYCLOAK_LAST_NAME = window.REACT_APP_KEYCLOAK_LAST_NAME
export const FEATURE_FLAGS = window.REACT_APP_FEATURE_FLAGS || []

// Maximum doses allowed by vaccine manufacturer
export const MAX_DOSES = window.REACT_APP_MAX_DOSES || {}

export const CONSENT_LABEL = window.REACT_APP_CONSENT_LABEL
export const ALTERNATE_CONSENT_LABEL = window.REACT_APP_ALTERNATE_CONSENT_LABEL

// Global country config
export const GLOBAL_CONFIG = window.REACT_APP_GLOBAL_CONFIG || {};

