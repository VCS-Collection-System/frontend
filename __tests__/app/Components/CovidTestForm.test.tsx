import React from 'react';
import reactRouterDom from 'react-router-dom'
import { render, fireEvent, act } from '@testing-library/react'
import { screen } from '@testing-library/dom'
import { shallow } from 'enzyme';
import CovidTestForm from '../../../src/app/Components/CovidTestForm';
import mockKeycloak from "../../../__mocks__/keycloakMock"
import { getConfig } from "../../../__mocks__/configMock"
import { CONFIG_URL } from "@app/utils/constants"

jest.mock("@react-keycloak/web", () => ({
  useKeycloak: jest.fn(() => ({ keycloak: mockKeycloak })),
}))

jest.mock("@app/utils/constants", () => ({
  FEATURE_FLAGS: [
    {
      "name": "consentCheckbox",
      "active": true
    }
  ],
  KEYCLOAK_COUNTRY: "attributes.agency[0]",
  KEYCLOAK_EMAIL: "email",
  KEYCLOAK_FIRST_NAME: "firstName",
  KEYCLOAK_LAST_NAME: "lastName",
  KEYCLOAK_ID: "attributes.workforceid[0]",
  MAX_DOSES: {"Pfizer": 2},
  CONFIG_URL: "config/"
}))

const config = {"default": {
    "allowPdf": false,
    "dateOfBirth": true,
    "consentCheckboxAlternate": false,
    "consentCheckbox": false,
    "covidTest": false,
    "customInput": true,
    "hrDashboard": true,
    "hrSearch": true,
    "cdc": false, 
    "divoc": false, 
    "greenPass": false,
    "recovery": false,
    "fakeFeature": true
  },
  "deu": {"consentCheckboxAlternate": true}
};

jest.mock('@app/utils/axiosInterceptor', () => ({
  get: (URL: string) => getConfig(URL, CONFIG_URL, config)
}))

jest.mock("react-router-dom");
const mockPush = jest.fn();
reactRouterDom.useHistory = jest.fn().mockReturnValue({push: mockPush});

describe('<CovidTestForm />', ()=> {
    it('should be defined', () => {
        expect(CovidTestForm).toBeDefined()
    })
    it("should render without crashing", () => {
        const wrapper = shallow(<CovidTestForm />)
        expect(wrapper)
    })
    it("should match existing snapshot", () => {
        const wrapper = shallow(<CovidTestForm />)
        expect(wrapper).toMatchSnapshot()
    })
    it("birth date should be after min birth date", async () => {
        render(<CovidTestForm />)
        await new Promise(process.nextTick);
        fireEvent.change(screen.getByLabelText("date_of_birth"), {target: {value: "12-31-1920"}});
        fireEvent.blur(screen.getByLabelText("date_of_birth"));
        expect(screen.getByLabelText("date_of_birth")).toHaveAttribute("aria-invalid", "true");
    })
    it("should be after min dose date", () => {
        render(<CovidTestForm />);
        fireEvent.change(screen.getByLabelText("test-results-date"), {target: {value: "06-01-2021"}});
        fireEvent.blur(screen.getByLabelText("test-results-date"));
        expect(screen.getByLabelText("test-results-date")).toHaveAttribute("aria-invalid", "true");
    })
    it("should accept without consent in non-consent countries", () => {
        render(<CovidTestForm />)
        expect(screen.queryByLabelText("consent-checkbox")).toBeFalsy();
    })
    it("should require consent in consent countries", async () => {
        mockKeycloak.profile.attributes.agency[0] = "DEU";
        render(<CovidTestForm />);
        await new Promise(process.nextTick);
        expect(screen.getByLabelText("consent-checkbox")).toBeTruthy();
        mockKeycloak.profile.attributes.agency[0] = "USA";
    })
    it("should be able to click Back to return to menu", () => {
        render(<CovidTestForm />);
        fireEvent.click(screen.getByText("Back"));
        expect(mockPush).toHaveBeenCalledWith('/attestmenu');
    })
})
