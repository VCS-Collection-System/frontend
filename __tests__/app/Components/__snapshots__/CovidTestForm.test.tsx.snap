// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CovidTestForm /> should match existing snapshot 1`] = `
<Form
  onSubmit={[Function]}
>
  <Feature
    name="consentCheckboxAlternate"
  >
    <Component
      control={
        Object {
          "_defaultValues": Object {},
          "_fields": Object {},
          "_formState": Object {
            "dirtyFields": Object {},
            "errors": Object {},
            "isDirty": false,
            "isSubmitSuccessful": false,
            "isSubmitted": false,
            "isSubmitting": false,
            "isValid": false,
            "isValidating": false,
            "submitCount": 0,
            "touchedFields": Object {},
          },
          "_formValues": Object {},
          "_getFieldArrayValue": [Function],
          "_getIsDirty": [Function],
          "_getWatch": [Function],
          "_isInAction": false,
          "_isMounted": false,
          "_names": Object {
            "array": Set {},
            "mount": Set {},
            "unMount": Set {},
            "watch": Set {},
          },
          "_proxyFormState": Object {
            "dirtyFields": false,
            "errors": false,
            "isDirty": false,
            "isValid": false,
            "isValidating": false,
            "touchedFields": false,
          },
          "_removeFields": [Function],
          "_shouldUnregister": undefined,
          "_subjects": Object {
            "array": ge {
              "observers": Array [],
            },
            "control": ge {
              "observers": Array [],
            },
            "state": ge {
              "observers": Array [],
            },
            "watch": ge {
              "observers": Array [],
            },
          },
          "_updateFieldArray": [Function],
          "_updateProps": [Function],
          "_updateValid": [Function],
          "_updateValues": [Function],
          "register": [Function],
          "unregister": [Function],
        }
      }
      name="consent"
      render={[Function]}
      rules={
        Object {
          "required": true,
        }
      }
    />
  </Feature>
  <FormGroup
    fieldId="employee-full-name"
    helperText="Make sure this matches your full name."
    label="Full Name"
  >
    <TextInput
      aria-label="employee-full-name"
      isDisabled={true}
      type="text"
    />
  </FormGroup>
  <Component
    control={
      Object {
        "_defaultValues": Object {},
        "_fields": Object {},
        "_formState": Object {
          "dirtyFields": Object {},
          "errors": Object {},
          "isDirty": false,
          "isSubmitSuccessful": false,
          "isSubmitted": false,
          "isSubmitting": false,
          "isValid": false,
          "isValidating": false,
          "submitCount": 0,
          "touchedFields": Object {},
        },
        "_formValues": Object {},
        "_getFieldArrayValue": [Function],
        "_getIsDirty": [Function],
        "_getWatch": [Function],
        "_isInAction": false,
        "_isMounted": false,
        "_names": Object {
          "array": Set {},
          "mount": Set {},
          "unMount": Set {},
          "watch": Set {},
        },
        "_proxyFormState": Object {
          "dirtyFields": false,
          "errors": false,
          "isDirty": false,
          "isValid": false,
          "isValidating": false,
          "touchedFields": false,
        },
        "_removeFields": [Function],
        "_shouldUnregister": undefined,
        "_subjects": Object {
          "array": ge {
            "observers": Array [],
          },
          "control": ge {
            "observers": Array [],
          },
          "state": ge {
            "observers": Array [],
          },
          "watch": ge {
            "observers": Array [],
          },
        },
        "_updateFieldArray": [Function],
        "_updateProps": [Function],
        "_updateValid": [Function],
        "_updateValues": [Function],
        "register": [Function],
        "unregister": [Function],
      }
    }
    defaultValue=""
    name="alt_first_name"
    render={[Function]}
  />
  <Component
    control={
      Object {
        "_defaultValues": Object {},
        "_fields": Object {},
        "_formState": Object {
          "dirtyFields": Object {},
          "errors": Object {},
          "isDirty": false,
          "isSubmitSuccessful": false,
          "isSubmitted": false,
          "isSubmitting": false,
          "isValid": false,
          "isValidating": false,
          "submitCount": 0,
          "touchedFields": Object {},
        },
        "_formValues": Object {},
        "_getFieldArrayValue": [Function],
        "_getIsDirty": [Function],
        "_getWatch": [Function],
        "_isInAction": false,
        "_isMounted": false,
        "_names": Object {
          "array": Set {},
          "mount": Set {},
          "unMount": Set {},
          "watch": Set {},
        },
        "_proxyFormState": Object {
          "dirtyFields": false,
          "errors": false,
          "isDirty": false,
          "isValid": false,
          "isValidating": false,
          "touchedFields": false,
        },
        "_removeFields": [Function],
        "_shouldUnregister": undefined,
        "_subjects": Object {
          "array": ge {
            "observers": Array [],
          },
          "control": ge {
            "observers": Array [],
          },
          "state": ge {
            "observers": Array [],
          },
          "watch": ge {
            "observers": Array [],
          },
        },
        "_updateFieldArray": [Function],
        "_updateProps": [Function],
        "_updateValid": [Function],
        "_updateValues": [Function],
        "register": [Function],
        "unregister": [Function],
      }
    }
    defaultValue=""
    name="alt_last_name"
    render={[Function]}
  />
  <Feature
    name="dateOfBirth"
  >
    <Component
      control={
        Object {
          "_defaultValues": Object {},
          "_fields": Object {},
          "_formState": Object {
            "dirtyFields": Object {},
            "errors": Object {},
            "isDirty": false,
            "isSubmitSuccessful": false,
            "isSubmitted": false,
            "isSubmitting": false,
            "isValid": false,
            "isValidating": false,
            "submitCount": 0,
            "touchedFields": Object {},
          },
          "_formValues": Object {},
          "_getFieldArrayValue": [Function],
          "_getIsDirty": [Function],
          "_getWatch": [Function],
          "_isInAction": false,
          "_isMounted": false,
          "_names": Object {
            "array": Set {},
            "mount": Set {},
            "unMount": Set {},
            "watch": Set {},
          },
          "_proxyFormState": Object {
            "dirtyFields": false,
            "errors": false,
            "isDirty": false,
            "isValid": false,
            "isValidating": false,
            "touchedFields": false,
          },
          "_removeFields": [Function],
          "_shouldUnregister": undefined,
          "_subjects": Object {
            "array": ge {
              "observers": Array [],
            },
            "control": ge {
              "observers": Array [],
            },
            "state": ge {
              "observers": Array [],
            },
            "watch": ge {
              "observers": Array [],
            },
          },
          "_updateFieldArray": [Function],
          "_updateProps": [Function],
          "_updateValid": [Function],
          "_updateValues": [Function],
          "register": [Function],
          "unregister": [Function],
        }
      }
      defaultValue=""
      name="date_of_birth"
      render={[Function]}
      rules={
        Object {
          "pattern": /\\^\\(0\\[1-9\\]\\|1\\[0-2\\]\\)\\\\-\\(0\\[1-9\\]\\|1\\\\d\\|2\\\\d\\|3\\[01\\]\\)\\\\-\\(19\\|20\\)\\\\d\\{2\\}\\$/,
          "required": true,
          "validate": [Function],
        }
      }
    />
  </Feature>
  <Component
    control={
      Object {
        "_defaultValues": Object {},
        "_fields": Object {},
        "_formState": Object {
          "dirtyFields": Object {},
          "errors": Object {},
          "isDirty": false,
          "isSubmitSuccessful": false,
          "isSubmitted": false,
          "isSubmitting": false,
          "isValid": false,
          "isValidating": false,
          "submitCount": 0,
          "touchedFields": Object {},
        },
        "_formValues": Object {},
        "_getFieldArrayValue": [Function],
        "_getIsDirty": [Function],
        "_getWatch": [Function],
        "_isInAction": false,
        "_isMounted": false,
        "_names": Object {
          "array": Set {},
          "mount": Set {},
          "unMount": Set {},
          "watch": Set {},
        },
        "_proxyFormState": Object {
          "dirtyFields": false,
          "errors": false,
          "isDirty": false,
          "isValid": false,
          "isValidating": false,
          "touchedFields": false,
        },
        "_removeFields": [Function],
        "_shouldUnregister": undefined,
        "_subjects": Object {
          "array": ge {
            "observers": Array [],
          },
          "control": ge {
            "observers": Array [],
          },
          "state": ge {
            "observers": Array [],
          },
          "watch": ge {
            "observers": Array [],
          },
        },
        "_updateFieldArray": [Function],
        "_updateProps": [Function],
        "_updateValid": [Function],
        "_updateValues": [Function],
        "register": [Function],
        "unregister": [Function],
      }
    }
    defaultValue=""
    name="test_results"
    render={[Function]}
    rules={
      Object {
        "required": true,
      }
    }
  />
  <Component
    control={
      Object {
        "_defaultValues": Object {},
        "_fields": Object {},
        "_formState": Object {
          "dirtyFields": Object {},
          "errors": Object {},
          "isDirty": false,
          "isSubmitSuccessful": false,
          "isSubmitted": false,
          "isSubmitting": false,
          "isValid": false,
          "isValidating": false,
          "submitCount": 0,
          "touchedFields": Object {},
        },
        "_formValues": Object {},
        "_getFieldArrayValue": [Function],
        "_getIsDirty": [Function],
        "_getWatch": [Function],
        "_isInAction": false,
        "_isMounted": false,
        "_names": Object {
          "array": Set {},
          "mount": Set {},
          "unMount": Set {},
          "watch": Set {},
        },
        "_proxyFormState": Object {
          "dirtyFields": false,
          "errors": false,
          "isDirty": false,
          "isValid": false,
          "isValidating": false,
          "touchedFields": false,
        },
        "_removeFields": [Function],
        "_shouldUnregister": undefined,
        "_subjects": Object {
          "array": ge {
            "observers": Array [],
          },
          "control": ge {
            "observers": Array [],
          },
          "state": ge {
            "observers": Array [],
          },
          "watch": ge {
            "observers": Array [],
          },
        },
        "_updateFieldArray": [Function],
        "_updateProps": [Function],
        "_updateValid": [Function],
        "_updateValues": [Function],
        "register": [Function],
        "unregister": [Function],
      }
    }
    defaultValue=""
    name="test_date"
    render={[Function]}
    rules={
      Object {
        "pattern": /\\^\\(0\\[1-9\\]\\|1\\[0-2\\]\\)\\\\-\\(0\\[1-9\\]\\|1\\\\d\\|2\\\\d\\|3\\[01\\]\\)\\\\-\\(19\\|20\\)\\\\d\\{2\\}\\$/,
        "required": true,
        "validate": [Function],
      }
    }
  />
  <Component
    control={
      Object {
        "_defaultValues": Object {},
        "_fields": Object {},
        "_formState": Object {
          "dirtyFields": Object {},
          "errors": Object {},
          "isDirty": false,
          "isSubmitSuccessful": false,
          "isSubmitted": false,
          "isSubmitting": false,
          "isValid": false,
          "isValidating": false,
          "submitCount": 0,
          "touchedFields": Object {},
        },
        "_formValues": Object {},
        "_getFieldArrayValue": [Function],
        "_getIsDirty": [Function],
        "_getWatch": [Function],
        "_isInAction": false,
        "_isMounted": false,
        "_names": Object {
          "array": Set {},
          "mount": Set {},
          "unMount": Set {},
          "watch": Set {},
        },
        "_proxyFormState": Object {
          "dirtyFields": false,
          "errors": false,
          "isDirty": false,
          "isValid": false,
          "isValidating": false,
          "touchedFields": false,
        },
        "_removeFields": [Function],
        "_shouldUnregister": undefined,
        "_subjects": Object {
          "array": ge {
            "observers": Array [],
          },
          "control": ge {
            "observers": Array [],
          },
          "state": ge {
            "observers": Array [],
          },
          "watch": ge {
            "observers": Array [],
          },
        },
        "_updateFieldArray": [Function],
        "_updateProps": [Function],
        "_updateValid": [Function],
        "_updateValues": [Function],
        "register": [Function],
        "unregister": [Function],
      }
    }
    name="test_image_file"
    render={[Function]}
    rules={
      Object {
        "required": true,
      }
    }
  />
  <FormGroup
    fieldId="accept-test-buttons"
  >
    <Button
      icon={
        <CheckCircleIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
        />
      }
      type="submit"
      variant="primary"
    >
      Accept
    </Button>
    <Button
      onClick={[Function]}
      variant="plain"
    >
      Back
    </Button>
  </FormGroup>
</Form>
`;
